javalib/src/main/scala/java/lang/String.scala

```diff
@@ -667,22 +668,259 @@ final class _String()
+        case '\u0049' if moreAbove(i) => "\u0069\u0307"
+        case '\u004A' if moreAbove(i) => "\u006A\u0307"
+        case '\u012E' if moreAbove(i) => "\u012F\u0307"
+        case '\u00CC'                 => "\u0069\u0307\u0300"
+        case '\u00CD'                 => "\u0069\u0307\u0301"
+        case '\u0128'                 => "\u0069\u0307\u0303"
+        case _                        => null
+      }
+    }
+
+    preprocessed.toLowerCase()
+  }
+
+  private def toLowerCaseTurkishAndAzeri(): String = {
+    /* Relevant excerpt from SpecialCasing.txt
+     *
+     * # Turkish and Azeri
+     *
+     * # I and i-dotless; I-dot and i are case pairs in Turkish and Azeri
+     * # The following rules handle those cases.
+     *
+     * 0130; 0069; 0130; 0130; tr; # LATIN CAPITAL LETTER I WITH DOT ABOVE
+     * 0130; 0069; 0130; 0130; az; # LATIN CAPITAL LETTER I WITH DOT ABOVE
+     *
+     * # When lowercasing, remove dot_above in the sequence I + dot_above, which will turn into i.
+     * # This matches the behavior of the canonically equivalent I-dot_above
+     *
+     * 0307; ; 0307; 0307; tr After_I; # COMBINING DOT ABOVE
+     * 0307; ; 0307; 0307; az After_I; # COMBINING DOT ABOVE
+     *
+     * # When lowercasing, unless an I is before a dot_above, it turns into a dotless i.
+     *
+     * 0049; 0131; 0049; 0049; tr Not_Before_Dot; # LATIN CAPITAL LETTER I
+     * 0049; 0131; 0049; 0049; az Not_Before_Dot; # LATIN CAPITAL LETTER I
+     */
+
+    /* Tests whether we are in an `After_I` context.
+     * From Table 3.17 in the Unicode standard:
+     * - Description: There is an uppercase I before C, and there is no
+     *   intervening combining character class 230 (Above) or 0.
+     * - Regex, before C: [I]([^\p{ccc=230}\p{ccc=0}])*
+     */
+    def afterI(i: Int): scala.Boolean = {
+      val j = skipCharsWithCombiningClassOtherThanNoneOrAboveBackwards(i)
+      j > 0 && charAt(j - 1) == 'I'
+    }
+
+    /* Tests whether we are in an `Before_Dot` context.
+     * From Table 3.17 in the Unicode standard:
+     * - Description: C is followed by combining dot above (U+0307). Any
+     *   sequence of characters with a combining class that is neither 0 nor
+     *   230 may intervene between the current character and the combining dot
+     *   above.
+     * - Regex, after C: ([^\p{ccc=230}\p{ccc=0}])*[\u0307]
+     */
+    def beforeDot(i: Int): scala.Boolean = {
+      val j = skipCharsWithCombiningClassOtherThanNoneOrAboveForwards(i + 1)
+      j != length() && charAt(j) == '\u0307'
+    }
+
+    val preprocessed = replaceCharsAtIndex { i =>
+      (this.charAt(i): @switch) match {
+        case '\u0130'                  => "\u0069"
+        case '\u0307' if afterI(i)     => ""
+        case '\u0049' if !beforeDot(i) => "\u0131"
+        case _                         => null
+      }
+    }
+
+    preprocessed.toLowerCase()
+  }
+
+  @inline def toLowerCase(): _String = toCase(Character.toLowerCase)
```

Comment:
To reduce the overall diff:
```suggestion
  @inline
  def toLowerCase(): _String = toCase(Character.toLowerCase)
```